### PR TITLE
[usbdev] Bring usbdev to D2(S)

### PR DESCRIPTION
**Pull request by a-will**
1. Add SVAs for validating parameters.
2. Clean up TODOs.
3. Update docs and add recipe for handling low power modes.
4. Update checklist.

### CODE REVIEW

File: hw/ip/usbdev/doc/wake_resume.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+---
+title: "Suspending and Resuming with Sleep / Low Power Modes"
+---
```
Comment (tjaychen) [inline hw/ip/usbdev/doc/wake_resume.md:3]: i might be slightly blind, but does this need to be linked in the main markdown anywhere?

File: hw/ip/usbdev/doc/_index.md [hunk 2]
```diff
@@ -251,8 +218,15 @@ In addition, the buffer ID, the packet size, an out/setup flag and the endpoint
 Software should immediately provide a free buffer for future reception by writing the corresponding buffer ID to the Available Buffer FIFO.
 It can then process the packet and eventually return the received buffer to the free pool.
 This allows streaming on a single endpoint or across a number of endpoints.
-If the packets cannot be consumed at the rate they are received, software can implement selective flow control by disabling OUT or SETUP transactions for a particular endpoint, which will result in a request to that endpoint being NACKed (negative acknowledgment).
-In the unfortunate event that the Available Buffer FIFO is empty or the Received Buffer FIFO is full, all OUT and SETUP transactions are NACKed.
+If the packets cannot be consumed at the rate they are received, software can implement selective flow control by disabling OUT transactions for a particular endpoint, which will result in a request to that endpoint being NAKed (negative acknowledgment).
```
Comment (tjaychen) [inline hw/ip/usbdev/doc/_index.md:221]: this is different from the end point being completely disabled right? I recall there was a discussion on an interface that's disabled should not even respond.

File: hw/ip/usbdev/doc/wake_resume.md (hunk 1 above)
Comment (a-will) [reply hw/ip/usbdev/doc/wake_resume.md:3]: Might be, yeah -- I haven't become familiar with how we structure our docs.

Currently, it just creates a separate page underneath the usbdev section. But... the regref stuff doesn't look like it'll work :smile:

File: hw/ip/usbdev/doc/_index.md (hunk 2 above)
Comment (a-will) [reply hw/ip/usbdev/doc/_index.md:221]: That's correct. In truth, I probably should rename the `rxenable_out` register somehow... "Enable" / "Disable" isn't really the right verb... It's more like "Accept" / "Reject" (except "Stall" overrides, from a separate register).

Do you have a preference?

Comment (tjaychen) [reply hw/ip/usbdev/doc/_index.md:221]: ah there's no need to change, i was really just asking a clarifying question.

File: hw/ip/usbdev/doc/_index.md [hunk 3]
```diff
@@ -304,71 +278,117 @@ Initially these will be sent to device ID 0.
 When a Set Address request is received, the device ID received must be stored in the {{< regref "usbctrl.device_address" >}} register.
 Note that device 0 is used for the entire control transaction setting the new device ID, so writing the new ID to the register should not be done until the ACK for the Status stage has been received (see [USB 2.0 specification](https://www.usb.org/document-library/usb-20-specification)).
 
+The host will then issue additional control transfers to Endpoint 0 to configure the device, now to the device's configured address.
+In response to the Set Configuration request, software should set up the rest of the endpoints for that configuration, including configuring the flow control behavior for OUT endpoints via the {{< regref "set_nak_out" >}} register, configuring the endpoint type via the {{< regref "rxenable_setup" >}} register (for a control endpoint) and the {{< regref "out_iso" >}} and {{< regref "in_iso" >}} registers (for isochronous OUT and IN endpoints, respectively).
+Finally, software should enable the configured endpoints via the {{< regref "ep_out_enable" >}} and {{< regref "ep_in_enable" >}} registers.
```
Comment (tjaychen) [inline hw/ip/usbdev/doc/_index.md:283]: ah yes, do we document anywhere what happens if ep_*_enable is not set and someone tries to talk to it?
I think this is related to the point we're making above.

Comment (a-will) [reply hw/ip/usbdev/doc/_index.md:283]: It's in the register doc for the fields:

For ep_out_enable:
```
                  If the bit is clear, any SETUP or OUT packets sent to the endpoint will be ignored.
```

For ep_in_enable:
```
                  If the bit is clear then any IN packets sent to the endpoint will be ignored.
```

**Review by tjaychen — APPROVED** [review]
this all looks good to me. 

i now have a separate question on the pull-ups, but we'll have to answer that separately.

File: hw/ip/usbdev/doc/wake_resume.md (hunk 1 above)
Comment (a-will) [reply hw/ip/usbdev/doc/wake_resume.md:3]: Added an ability for regref to link to other pages, so this one links to the main USBDEV page.

**Comment by tjaychen** [comment]
any objections to merging?

**Pull request merged**